velox/exec/HashBuild.h

```diff
@@ -18,35 +18,103 @@
-// the same name.
+/// Hands over a hash table from a multi-threaded build pipeline to a
+/// multi-threaded probe pipeline. This is owned by shared_ptr by all the build
+/// and probe Operator instances concerned. Corresponds to the Presto concept of
+/// the same name.
 class HashJoinBridge : public JoinBridge {
  public:
-  void setHashTable(std::unique_ptr<BaseHashTable> table);
+  void start() override;
+
+  /// Invoked by HashBuild operator ctor to add to this bridge by incrementing
+  /// 'numBuilders_'. The latter is used to split the spill partition data among
+  /// HashBuild operators to parallelize the restoring operation.
+  void addBuilder();
+
+  /// 'spillPartitionSet' contains the spilled partitions while building
+  /// 'table'. The function returns true if there is spill data to restore
+  /// after HashProbe operators process 'table', otherwise false. This only
+  /// applies if the disk spilling is enabled.
+  bool setHashTable(
+      std::unique_ptr<BaseHashTable> table,
+      SpillPartitionSet spillPartitionSet);
 
   void setAntiJoinHasNullKeys();
 
-  // Represents the result of a HashBuild operator: a hash table. In case of an
-  // anti join, a build side entry with a null in a join key makes the join
-  // return nothing. In this case, HashBuild operator finishes early without
-  // processing all the input and without finishing building the hash table.
+  /// Represents the result of HashBuild operators: a hash table, an optional
+  /// restored spill partition id associated with the table, and the spilled
+  /// partitions while building the table if not empty. In case of an anti join,
+  /// a build side entry with a null in a join key makes the join return
+  /// nothing. In this case, HashBuild operators finishes early without
+  /// processing all the input and without finishing building the hash table.
   struct HashBuildResult {
+    HashBuildResult(
+        std::shared_ptr<BaseHashTable> _table,
+        std::optional<SpillPartitionId> _restoredPartitionId,
+        SpillPartitionIdSet _spillPartitionIds)
+        : antiJoinHasNullKeys(false),
+          table(std::move(_table)),
+          restoredPartitionId(std::move(_restoredPartitionId)),
+          spillPartitionIds(std::move(_spillPartitionIds)) {}
+
+    HashBuildResult() : antiJoinHasNullKeys(true) {}
+
+    bool antiJoinHasNullKeys{false};
     std::shared_ptr<BaseHashTable> table;
-    bool antiJoinHasNullKeys;
+    std::optional<SpillPartitionId> restoredPartitionId;
+    SpillPartitionIdSet spillPartitionIds;
+  };
+
+  std::optional<HashBuildResult> tableOrFuture(
+      ContinueFuture* FOLLY_NONNULL future);
+
+  /// Invoked by HashProbe operator after finishes probing the built table to
+  /// set one of the previously spilled partition to restore. The HashBuild
+  /// operators will then build the next hash table from the selected spilled
+  /// one.
+  bool setSpillInput();
```

Comment:
Yeah, probeFinished is better name and it will be called by the last probe operator. The actual wait for the next table is handled by the other API. probeFinished is definitely better than setSpillInput and I actually thought the name for a while. While document the semantics internally. Thanks for the suggestion!